src/main/java/sirius/biz/tenants/flags/CustomizationFlags.java

```diff
@@ -0,0 +1,137 @@
+ * <p>
+ * This is mainly intended for migration projects where functionality needs either to be enabled for specific users
+ * or scopes or needs to be "set back" to the old behavior. Therefore, a flag can be toggled in many places:
+ * <ol>
+ *     <li>By either having the string <tt>flag-FLAGNAME-disabled</tt> or <tt>flag-FLAGNAME</tt> in the user agent</li>
+ *     <li>By setting it in the custom config in the current user or its tenant (in the block <tt>flags</tt>)</li>
+ *     <li>By setting it in the scope config</li>
+ *     <li>By setting it in the tenant config which owns the current scope</li>
+ *     <li>By setting it in the config of the system tenant</li>
+ * </ol>
+ * <p>
+ * These places are checked in order and the first given value is used. Therefore, especially via the system tenant,
+ * a whole system can be set to either "all on" or "all off" - with specific exceptions on multiple levels.
+ * <p>
+ * Note that the API should be mainly accessed via {@link CustomizationFlag}.
+ */
+@Register(classes = CustomizationFlags.class)
+public class CustomizationFlags {
+
+    private static final List<String> knownFlags = new ArrayList<>();
+
+    @Part
+    private Tenants<?, ?, ?> tenants;
+
+    protected static synchronized void addKnownFlag(String name) {
+        knownFlags.add(name);
+    }
+
+    /**
+     * Reports a list of all statically known flags.
+     *
+     * @return a list fo all flags which are created via {@link CustomizationFlag}.
+     */
+    public synchronized List<String> getKnownFlags() {
+        return new ArrayList<>(knownFlags);
+    }
+
+    /**
+     * Checks if the given flag is enabled or disabled and also reports the source of the value.
+     * <p>
+     * Note that this is rather a debugging / reporting API and the main access should be performed via
+     * {@link CustomizationFlag#isEnabled()} or {@link CustomizationFlag#isEnabled(boolean)}. This way, all
+     * flags which are in use, get properly reported via {@link #getKnownFlags()}.
+     *
+     * @param flagName     the name of the flag to query
+     * @param defaultValue the default to use in case no config is present anywhere
+     * @return the value of the flag and the source where it was found
+     */
+    public Tuple<Boolean, String> isFlagEnabled(String flagName, boolean defaultValue) {
+
+        return readFromUserAgent(flagName).or(() -> readFromUserSettings(flagName))
+                                          .or(() -> readFromScopeSettings(flagName))
+                                          .or(() -> readFromScopeTenantSettings(flagName))
+                                          .or(() -> readFromSystemTenantSettings(flagName))
+                                          .orElseGet(() -> Tuple.create(defaultValue, "default"));
+    }
+
+    private Optional<Tuple<Boolean, String>> readFromUserAgent(String flagName) {
+        WebContext webContext = CallContext.getCurrent().get(WebContext.class);
```

Comment:
**Question:** Can this be null or fail when checking a flag in a background process or job?

Reply:
Nope - this would yiel and empty WebContext which is not valid (see below)